### PR TITLE
[SYCLomatic] Make sure thrust::sort_by_key is migrated correctly for USM

### PR DESCRIPTION
**Pull request by LU-JOHN**
Signed-off-by: Lu, John [john.lu@intel.com](mailto:john.lu@intel.com)

### CODE REVIEW

File: clang/test/dpct/thrust_sort_by_key_usm.cu [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ 3. Test each VERSION with (device_vector/host_vector/malloc-ed memory/cudaMalloc-ed memory)
+ *******************************************************************************************/
+
+/*********** sort_by_key ***********************************************************************************************************************************************/
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// CHECK:dpct::sort(oneapi::dpl::execution::seq, AH.begin(), AH.end(), BH.begin());
+// CHECK-NEXT:dpct::sort(oneapi::dpl::execution::make_device_policy(q_ct1), AD.begin(), AD.end(), BD.begin());
+// CHECK-NEXT:dpct::sort(oneapi::dpl::execution::seq, h_ptr, h_ptr + 4, BH.begin());
+// CHECK-NEXT:dpct::sort(oneapi::dpl::execution::seq, d_ptr, d_ptr + 4, BD.begin());
```
Comment (tomflinda) [inline clang/test/dpct/thrust_sort_by_key_usm.cu:38]: From line43, d_ptr, BD.begin() are device ptr and device iterator,  so the policy should be `oneapi::dpl::execution::make_device_policy(q_ct1)`.

File: clang/test/dpct/thrust_sort_by_key_usm.cu [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// CHECK:dpct::sort(oneapi::dpl::execution::seq, AH.begin(), AH.end(), BH.begin(), std::greater<int>());
+// CHECK-NEXT:dpct::sort(oneapi::dpl::execution::make_device_policy(q_ct1), AD.begin(), AD.end(), BD.begin(), std::greater<int>());
+// CHECK-NEXT:dpct::sort(oneapi::dpl::execution::seq, h_ptr, h_ptr + 4, BH.begin(), std::greater<int>());
+// CHECK-NEXT:dpct::sort(oneapi::dpl::execution::seq, d_ptr, d_ptr + 4, BD.begin(), std::greater<int>());
+  // VERSION                          first       last      result      comparator
+  thrust::sort_by_key(                AH.begin(), AH.end(), BH.begin(), thrust::greater<int>());
+  thrust::sort_by_key(                AD.begin(), AD.end(), BD.begin(), thrust::greater<int>());
+  thrust::sort_by_key(                h_ptr,      h_ptr+4,  BH.begin(), thrust::greater<int>());  
+  thrust::sort_by_key(                d_ptr,      d_ptr+4,  BD.begin(), thrust::greater<int>());
```
Comment (tomflinda) [inline clang/test/dpct/thrust_sort_by_key_usm.cu:54]: Same as above

File: clang/test/dpct/thrust_stable_sort_by_key_usm.cu [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// CHECK:dpct::stable_sort(oneapi::dpl::execution::seq, AH.begin(), AH.end(), BH.begin());
+// CHECK-NEXT:dpct::stable_sort(oneapi::dpl::execution::make_device_policy(q_ct1), AD.begin(), AD.end(), BD.begin());
+// CHECK-NEXT:dpct::stable_sort(oneapi::dpl::execution::seq, h_ptr, h_ptr + 4, BH.begin());
+// CHECK-NEXT:dpct::stable_sort(oneapi::dpl::execution::seq, d_ptr, d_ptr + 4, BD.begin());
+  // VERSION                                 first       last      result
+  thrust::stable_sort_by_key(                AH.begin(), AH.end(), BH.begin());
+  thrust::stable_sort_by_key(                AD.begin(), AD.end(), BD.begin());
+  thrust::stable_sort_by_key(                h_ptr,      h_ptr+4,  BH.begin());
+  thrust::stable_sort_by_key(                d_ptr,      d_ptr+4,  BD.begin());
```
Comment (tomflinda) [inline clang/test/dpct/thrust_stable_sort_by_key_usm.cu:43]: Same as before

File: clang/test/dpct/thrust_stable_sort_by_key_usm.cu [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// CHECK:dpct::stable_sort(oneapi::dpl::execution::seq, AH.begin(), AH.end(), BH.begin(), std::greater<int>());
+// CHECK-NEXT:dpct::stable_sort(oneapi::dpl::execution::make_device_policy(q_ct1), AD.begin(), AD.end(), BD.begin(), std::greater<int>());
+// CHECK-NEXT:dpct::stable_sort(oneapi::dpl::execution::seq, h_ptr, h_ptr + 4, BH.begin(), std::greater<int>());
+// CHECK-NEXT:dpct::stable_sort(oneapi::dpl::execution::seq, d_ptr, d_ptr + 4, BD.begin(), std::greater<int>());
+  // VERSION                                 first       last      result      comparator
+  thrust::stable_sort_by_key(                AH.begin(), AH.end(), BH.begin(), thrust::greater<int>());
+  thrust::stable_sort_by_key(                AD.begin(), AD.end(), BD.begin(), thrust::greater<int>());
+  thrust::stable_sort_by_key(                h_ptr,      h_ptr+4,  BH.begin(), thrust::greater<int>());  
+  thrust::stable_sort_by_key(                d_ptr,      d_ptr+4,  BD.begin(), thrust::greater<int>());
```
Comment (tomflinda) [inline clang/test/dpct/thrust_stable_sort_by_key_usm.cu:54]: Same as before.

File: clang/test/dpct/thrust_sort_by_key_usm.cu (hunk 1 above)
Comment (LU-JOHN) [reply clang/test/dpct/thrust_sort_by_key_usm.cu:38]: SYCLomatic cannot differentiate between h_ptr and d_ptr without data-flow analysis.  The types of h_ptr and d_ptr are both "int *".    The execution policy is chosen by the test:

   CheckThrustArgType(1, "thrust::device_ptr")
   
 The same problem occurs with other functions.  For example:
 
     void foo() {
      thrust::device_vector<int> BD(4);
      int *d_ptr;
      cudaMalloc(&d_ptr, 20 * sizeof(int));
      thrust::unique_by_key(d_ptr,      d_ptr+4,  BD.begin());
    }

Migrates to:

    void foo() {
      dpct::device_vector<int> BD(4);
      int *d_ptr;
      d_ptr = sycl::malloc_device<int>(20, dpct::get_default_queue());
      dpct::unique(oneapi::dpl::execution::seq, d_ptr, d_ptr + 4, BD.begin());
    }

Comment (tomflinda) [reply clang/test/dpct/thrust_sort_by_key_usm.cu:38]: For USM case, SYCL provides get_pointer_type() to identify the type of the pointer. We have two options:
1. Generate a if stmt in the migrated code, just like what we did in non-USM;
2. Create a helper function and put the if stmt in the helper function.

Or we can merge the current PR and leave the rest work for future improvement.

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**